includes/class-sensei-course.php

```diff
@@ -3068,40 +3069,40 @@ public static function is_prerequisite_complete( $course_id ){
     }// end is_prerequisite_complete
 
 	/**
-	 * Allowing user to set course archive page as front page
+	 * Allowing user to set course archive page as front page.
+	 *
+	 * Expects to be called during pre_get_posts, but only if page_on_front is set
+	 * to a non-empty value.
 	 *
 	 * @since 1.9.5
 	 * @param WP_Query $query hooked in from pre_get_posts
 	 */
 	function allow_course_archive_on_front_page( $query ) {
+		// Flag used to indicate if our test query is running: we use this instead of temporarily
+		// hooking/unhooking due to a core WP bug @see https://core.trac.wordpress.org/ticket/17817
+		static $running = false;
 
-		global $wp;
-
-		if ( ! is_page() ){
-			return; // This only needs to run on the home page. See https://github.com/Automattic/sensei/issues/1438
+		// Bail if it's clear we're not looking at a static front page or if the $running flag is
+		// set @see https://github.com/Automattic/sensei/issues/1438
+		if ( $running || ! $query->is_main_query() || ! is_page() || is_admin() ) {
```

Comment:
Is this the most efficient order of operations for the tests?  I'd sooner have the cheapest, most likely to short-circuit tests put first to optimize.